include/rocksdb/options.h

```diff
@@ -78,6 +78,50 @@ enum CompressionType : unsigned char {
 
 struct Options;
 struct DbPath;
+enum DbPathPlacementStrategy : unsigned char {
+  // Fill DbPaths according to the target size set with the db path.
+  //
+  // Newer data is placed into paths specified earlier in the vector while
+  // older data gradually moves to paths specified later in the vector.
+  //
+  // For example, you have a flash device with 10GB allocated for the DB,
+  // as well as a hard drive of 2TB, you should config it to be:
+  //   [{"/flash_path", 10GB}, {"/hard_drive", 2TB}]
+  //
+  // The system will try to guarantee data under each path is close to but
+  // not larger than the target size. But current and future file sizes used
+  // by determining where to place a file are based on best-effort estimation,
+  // which means there is a chance that the actual size under the directory
+  // is slightly more than target size under some workloads. User should give
+  // some buffer room for those cases.
+  //
+  // If none of the paths has sufficient room to place a file, the file will
+  // be placed to the last path anyway, despite to the target size.
+  //
+  // Placing newer data to earlier paths is also best-efforts. User should
+  // expect user files to be placed in higher levels in some extreme cases.
+  kGradualMoveOldDataTowardsEnd = 0x00,
+
+  // Randomly distribute files into the list of db paths.
+  //
+  // For example, you have a few data drives on your host that are mounted
+  // as [/sdb1, /sdc1, /sdd1, /sde1]. Say that the database will create 6
+  // table files -- 0[0-5].sst, then they will end up on in these places:
+  //
+  // /sdb1/02.sst
+  // /sdb1/04.sst
+  // /sdc1/05.sst
+  // /sdc1/03.sst
+  // /sdd1/00.sst
+  // /sde1/01.sst
+  //
+  // This is useful if you want the database to evenly use a set of disks
+  // mounted on your host.
+  //
+  // Note that the target_size attr in DbPath will not be useful if this
+  // strategy is chosen.
+  kRandomlyChoosePath = 0x01,
+};
```

Comment:
Do you consider to make DbPathSupplier a public interface, and options take a factory class for it? If the factory is empty, then we fall back to the build-in behavior, otherwise, users can implement it in a way they want?

It can be done just like FlushBlockPolicy and FlushBlockBySizePolicyFactory. We have similar data structures like that.

Reply:
I really like this idea and I can see ourselves benefit from this more flexible approach. The ability to expand functionality without sending PR or forking Rocksdb is great.

The following reasons, however, are why I'm not already doing it:
1. I don't know how should we deal with the Java interface, which we plan to use in our application
2. I'm not entirely sure how should the factory interface look like. The current use cases will at least require the factory to have a method 
```cpp
virtual std::unique_ptr<DbPathSupplier> produce(
    const ImmutableOptions& ioptions, 
    const MutableOptions& moptions, 
    uint64_t estimated_file_size)
```
The `estimated_file_size` comes from the universal compaction path choosing logic. The output size is estimated upfront before the compaction actually happens (where `GetPathId()` is called on the supplier). This does feel a bit awkward because `estimated_file_size` is only useful for universal compaction but is added everywhere. There maybe another use case that requires another argument on the factory, thus breaking client side code.

Do you have suggestions for the above confusions?